pull request for issue #162

hacitandogan opened the pull request:
Fixes #162 

Changes proposed in this pull request:
clearing the duplicates from MiscPssdiaginfo.sql  and SQL Server Perf Stats Snapshot.sql and combining all AG related queries under AlwaysOnDiagScript.sql

How to test this code:
- execute T-SQL directly against AG instances
- build the pssdiag.zip and capture data from different AG versions
- import output of this into Nexus

Has been tested on (remove any that don't apply):
 - Case-sensitive SQL Server instance
 - SQL Server 2012
 - SQL Server 2014
 - SQL Server 2016
 - SQL Server 2017
 - SQL Server 2019

PiJoCoder commented on DiagManager/CustomDiagnostics/AlwaysOn Basic Info/AlwaysOnDiagScript.sql:
```diff
@@ -8,32 +8,48 @@ PRINT '==========================='
 PRINT 'Database Mirroring Endpoint'
 PRINT '==========================='
 PRINT ''
+PRINT '--AG_hadr_endpoints_principals--'
```
Can you add spaces between the dashes and the name - just for consistency with most identifiers. Trying to get them all to be using spaces:

```suggestion
PRINT '-- AG_hadr_endpoints_principals --'
```

PiJoCoder commented on DiagManager/CustomDiagnostics/AlwaysOn Basic Info/AlwaysOnDiagScript.sql:
```diff
@@ -8,32 +8,48 @@ PRINT '==========================='
-WHERE perm.class_desc = 'ENDPOINT' AND
-perm.permission_name = 'CONNECT' AND
-tep.type = 4
+PRINT '--AG_mirroring_endpoints_permissions--'
```
```suggestion
PRINT '-- AG_mirroring_endpoints_permissions --'
```

PiJoCoder commented on DiagManager/CustomDiagnostics/AlwaysOn Basic Info/AlwaysOnDiagScript.sql:
```diff
@@ -42,37 +58,55 @@ PRINT '==================================='
 PRINT 'Availability Group Listeners and IP'
 PRINT '==================================='
 PRINT ''
+PRINT '--AG_hadr_ag_listeners--'
```
```suggestion
PRINT '-- AG_hadr_ag_listeners --'
```

PiJoCoder commented on DiagManager/CustomDiagnostics/AlwaysOn Basic Info/AlwaysOnDiagScript.sql:
```diff
@@ -81,7 +115,9 @@ PRINT '========================'
 PRINT '========================'
 PRINT ''
-SELECT  *
+PRINT '--AG_hadr_cluster--'
```
```suggestion
PRINT '-- AG_hadr_cluster --'
```

PiJoCoder commented on DiagManager/CustomDiagnostics/AlwaysOn Basic Info/AlwaysOnDiagScript.sql:
```diff
@@ -92,41 +128,84 @@ PRINT '================================================'
-	inner join sys.dm_hadr_cluster_networks cn
-		on cn.member_name = cm.member_name
-
+PRINT '--AG_hadr_cluster_members--'
```
```suggestion
PRINT '-- AG_hadr_cluster_members --'
```

PiJoCoder commented on DiagManager/CustomDiagnostics/AlwaysOn Basic Info/AlwaysOnDiagScript.sql:
```diff
@@ -92,41 +128,84 @@ PRINT '================================================'
+		 ags.secondary_recovery_health_desc, ags.synchronization_health, ags.synchronization_health_desc
+	  FROM	sys.availability_groups AS ag 
+INNER JOIN	sys.dm_hadr_availability_group_states AS ags ON ag.group_id = ags.group_id'
+PRINT '--AG_hadr_ag_states--'
```
```suggestion
PRINT '-- AG_hadr_ag_states --'
```

PiJoCoder commented on DiagManager/CustomDiagnostics/AlwaysOn Basic Info/AlwaysOnDiagScript.sql:
```diff
@@ -135,20 +214,63 @@ PRINT '=========================================================================
 
-PRINT '-> dm_os_server_diagnostics_log_configurations'
-select * from sys.dm_os_server_diagnostics_log_configurations
+PRINT '--AG_dm_os_server_diagnostics_log_configurations--'
```
```suggestion
PRINT '-- AG_dm_os_server_diagnostics_log_configurations --'
```

PiJoCoder commented on DiagManager/CustomDiagnostics/AlwaysOn Basic Info/AlwaysOnDiagScript.sql:
```diff
@@ -92,41 +128,84 @@ PRINT '================================================'
+INNER JOIN	sys.availability_replicas                       AS ar   ON ars.replica_id  = ar.replica_id 
+INNER JOIN	sys.availability_groups                         AS ag   ON ag.group_id     = arcs.group_id AND ag.name = arc.group_name
+ORDER BY CAST(arc.group_name AS varchar(30)), CAST(ars.role_desc AS varchar(30))'
+PRINT '--AG_hadr_ag_replica_states--'
```
```suggestion
PRINT '-- AG_hadr_ag_replica_states --'
```

PiJoCoder commented on DiagManager/CustomDiagnostics/AlwaysOn Basic Info/AlwaysOnDiagScript.sql:
```diff
@@ -42,37 +58,55 @@ PRINT '==================================='
+
+SET @sql ='SELECT agl.dns_name AS [Listener_Name], ag.name AS [AG_Name] ,agl.group_id,listener_id,dns_name,port,is_conformant,ip_configuration_string_from_cluster'
+
+IF (@sql_major_version >=14) --this exists SQL 2017 and above
+	BEGIN
+	  SET @sql = @sql + ',is_distributed_network_name'
+	END
```
I ran the script on a SQL 2019 and got this error:
```
-- AG_hadr_ag_listeners --
Msg 207, Level 16, State 1, Server NODE1LABCSS\SQL2019, Line 1
Invalid column name 'is_distributed_network_name'.

Then ran `select * from sys.availability_group_listeners` and don't see this column. The suggestion is to remove this logic or double-check it and make sure it applies to this DMV

```
```suggestion

```

hacitandogan replied:
Thank you @PiJoCoder , I see this is not there on the RTM and  added later in the CUs and not failed for me since I have those CUs on my test box. I will modify the logic to check the builds as well per major version. 
https://docs.microsoft.com/en-us/azure/azure-sql/virtual-machines/windows/availability-group-distributed-network-name-dnn-listener-configure?view=azuresql

PiJoCoder replied:
Oh, I see, it is the SYS.AVAILABILITY_GROUP_LISTENERS DMV that has this.  I do have it on my SQL 2019 build, but was querying the sys.availability_groups instead. And since the documentation does not include the column, I didn't know which one had it. Can you fully-qualify the column to help with this discovery - ag1.is_distributed_network_name? And perhaps for consistency these columns too: listener_id,dns_name,port,is_conformant,ip_configuration_string_from_cluster
Thank you

hacitandogan replied:
@PiJoCoder , thank you , I modified this to make names full-qualified for those columns and added the additional build check logic for is_distributed_network_name column. 
--this column exists  SQL 2019 CU8 ,SQL 2017 CU25,SQL 2016 SP3 and above

PiJoCoder replied:
--this column exists SQL 2019 CU8 ,SQL 2017 CU25,SQL 2016 SP3 and above
Thanks for calling these out. Based on these, submitted an update request on 
https://docs.microsoft.com/en-us/sql/relational-databases/system-catalog-views/sys-availability-group-listeners-transact-sql?view=sql-server-ver16

PiJoCoder commented:
@hacitandogan , could you also sync your forked repo with recent changes from the upstream repo. Go to Fetch Upstream and sync your Forked repo. Then pull the changes down to your local machine (git pull)

![image](https://user-images.githubusercontent.com/23519517/175168054-76855f99-939c-4bcb-ab69-62866c215e59.png)

![image](https://user-images.githubusercontent.com/23519517/175168277-0bfd7b0c-f899-43bf-b916-8ed06a846f2f.png)

hacitandogan commented:
@PiJoCoder  , thank you for all the feedbacks.

I did following modifications;
-added build check for  is_distributed_network_name column
-added missing identifiers for xevent data
-removed headings
-other minor modifications

PiJoCoder commented:
@hacitandogan  one thing that I am discovering - in AG_AlwaysOn_health_availability_replica_manager_state_change output there could be a ton of rows. In fact the AG I just tested against returned over 5200 rows. Do you think we should get the latest 500 or something like that? Some limited number....?

hacitandogan commented:
Thank you @PiJoCoder  , just checked these numbers on my side and I see similar row numbers even on my test box.
To have a logical limit for these row counts, I am adding TOP 500 to all AlwaysOn_health* ones and changing date order to DESC to get the latest records. 
I hope this is fine.

PiJoCoder commented:
> Thank you @PiJoCoder , just checked these numbers on my side and I see similar row numbers even on my test box. To have a logical limit for these row counts, I am adding TOP 500 to all AlwaysOn_health* ones and changing date order to DESC to get the latest records. I hope this is fine.

Yes, this is good. At least we won't get a ton of rows that we don't need and bloat the output files.
Thank you. I will test and let you know if there's more feedback. If it looks good I'll merge into Master. I have assigned the SQL Nexus issue to you if you would like to update the Textrowset.xml file for these new items.

PiJoCoder commented:
Thank you @hacitandogan ! Looks great - merged!
